### PR TITLE
Adding tests for osa container images.

### PR DESCRIPTION
**Pull request by kwoodson**
Was recently an issue (#429) but moved to a feature https://jira.coreos.com/browse/AZURE-153

### CODE REVIEW

**Comment by jim-minter** [comment]
big apologies, I'm having difficulty remembering the rationale behind either of these requested tests now, or understanding their validity.  Specifically what things are we looking for here that can't be tested via unit testing @mjudeikis @kwoodson ?

**Comment by mjudeikis** [comment]
> I'm having difficulty remembering the rationale behind either of these requested tests now, or understanding their validity. Specifically what things are we looking for here that can't be tested via unit testing 

Because we had a case were we published an image with dev credentials as a valid image or centos image with rhel config. To prevent mistakes like this in the future. especially when we changing our node build process. 

We cant test this in the unit test because of its VM image specific test case. So Azure e2e is the place

File: test/e2e/osa_test.go [hunk 1]
```diff
@@ -83,4 +84,43 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
+		}
+
+		pods, err := c.kc.CoreV1().Pods("").List(metav1.ListOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		for _, pod := range pods.Items {
+			// e2e check should ensure that no reg-aws images are running on box
+			Expect(strings.Contains(pod.Status.ContainerStatuses[0].Image, "registry.reg-aws")).ToNot(BeTrue())
+			parts := strings.Split(pod.Status.ContainerStatuses[0].Image, "/")
+
+			if len(parts) < 3 {
```
Comment (kwoodson) [inline test/e2e/osa_test.go]: @mjudeikis, would it be better to loop through the pods in my `podsToCheck` map and ensure their registry is correct?

If by some reason their images don't align with this structure:
```
repo/namespace/container
```
then we skip the test as the image format is unknown.  

Let me experiment with this.

File: test/e2e/osa_test.go [hunk 2]
```diff
@@ -83,4 +84,43 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
+	It("should run the correct image", func() {
+		// image format is either:
+		// docker.io/openshift || registry.access.redhat.com/openshift3
+		format := strings.Split(c.cs.Config.Images.Format, "/")
+		prefix := fmt.Sprintf("%v/%v", format[0], format[1])
+
+		// these pods should have the same prefix
+		podsToCheck := map[string]struct{}{
+			"oauth-proxy":                    {},
+			"origin-control-plane":           {},
```
Comment (kargakis) [inline test/e2e/osa_test.go]: The suffix for RHEL is `ose-control-plane`. Similar I guess about all other origin instances.

Comment (kwoodson) [reply test/e2e/osa_test.go]: Thanks for pointing this out.  I am currently reworking the test and will attempt to determine ose vs origin.

**Comment by kwoodson** [comment]
After a discussion with @mjudeikis, we decided to simplify this check.  The issue that was raised has been addressed by the following:
1. Ensure that all pods are not running `reg-aws` registry.
2. Ensure that the node configs match the image format specified.

File: test/e2e/osa_test.go [hunk 3]
```diff
@@ -83,4 +83,24 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
 		Expect(err).ToNot(HaveOccurred())
 	})
+
+	It("should run the correct image", func() {
+		pods, err := c.kc.CoreV1().Pods("").List(metav1.ListOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		// e2e check should ensure that no reg-aws images are running on box
+		for _, pod := range pods.Items {
+			for _, cntr := range pod.Spec.Containers {
+				Expect(strings.Contains(cntr.Image, "registry.reg-aws")).ToNot(BeTrue())
```
Comment (jim-minter) [inline test/e2e/osa_test.go]: can you use HasPrefix?

File: test/e2e/osa_test.go [hunk 4]
```diff
@@ -83,4 +83,24 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
+			}
+		}
+
+		// Check all Configmaps for image format matches distribution
+		// format: registry.access.redhat.com/openshift3/ose-${component}:${version}
+		maps, err := c.kc.CoreV1().ConfigMaps("openshift-node").List(metav1.ListOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		for _, cm := range maps.Items {
+			Expect(strings.Contains(cm.Data["node-config.yaml"], "format: "+c.cs.Config.Images.Format)).To(BeTrue())
```
Comment (jim-minter) [inline test/e2e/osa_test.go]: As this stands I think this reduces to a unit test.
I think you should be determining the cluster OS (centos or rhel) by inspecting the Node object, then calculating what you expect the imageformat to be based on that, then testing it

File: test/e2e/osa_test.go (hunk 3 above)
Comment (kwoodson) [reply test/e2e/osa_test.go]: Done.

File: test/e2e/osa_test.go (hunk 4 above)
Comment (kwoodson) [reply test/e2e/osa_test.go]: Updated to check master-00000's os type and then test the image formats.

File: test/e2e/osa_test.go [hunk 5]
```diff
@@ -83,4 +83,35 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
 		Expect(err).ToNot(HaveOccurred())
 	})
+
+	It("should run the correct image", func() {
+		pods, err := c.kc.CoreV1().Pods("").List(metav1.ListOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		// e2e check should ensure that no reg-aws images are running on box
+		for _, pod := range pods.Items {
+			for _, cntr := range pod.Spec.Containers {
+				Expect(strings.HasPrefix(cntr.Image, "registry.reg-aws")).ToNot(BeTrue())
```
Comment (jim-minter) [inline test/e2e/osa_test.go]: "registry.reg-aws.openshift.com/" ?

File: test/e2e/osa_test.go [hunk 6]
```diff
@@ -83,4 +83,35 @@ var _ = Describe("Openshift on Azure e2e tests", func() {
...
+		} else {
+			registryPrefix = "docker.io/openshift"
+		}
+
+		// Check all Configmaps for image format matches master's OS type
+		// format: registry.access.redhat.com/openshift3/ose-${component}:${version}
+		maps, err := c.kc.CoreV1().ConfigMaps("openshift-node").List(metav1.ListOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		for _, cm := range maps.Items {
+			Expect(strings.Contains(cm.Data["node-config.yaml"], "format: "+registryPrefix)).To(BeTrue())
```
Comment (jim-minter) [inline test/e2e/osa_test.go]: Hacky - better to parse the yaml and check it properly

**Comment by kwoodson** [comment]
Tested on both origin and ose clusters.

**Comment by kwoodson** [comment]
Rebased to resolve conflicts.

**Comment by kwoodson** [comment]
Fixed rebase imports.

**Comment by mjudeikis** [comment]
I think I good with it, if @jim-minter is fine too - squash and merge?

**Comment by kargakis** [comment]
lgtm

**Pull request merged**